internal/topology/check/compatibility.go

```diff
@@ -176,10 +177,153 @@ func LocalObjectTemplateIsValid(in *clusterv1.LocalObjectTemplate, namespace str
 		allErrs = append(allErrs,
 			field.Invalid(
 				pathPrefix.Child("ref", "apiVersion"),
-				in.Ref.APIVersion,
+				template.Ref.APIVersion,
 				"value cannot be empty",
 			),
 		)
 	}
 	return allErrs
 }
+
+// ClusterClassesAreCompatible checks the compatibility between new and old versions of a Cluster Class.
+// It checks that:
+// 1) Infrastructure Templates are compatible.
+// 2) ControlPlane Templates are compatible.
+// 3) ControlPlane Machine Infrastructure Templates are compatible.
+// 4) MachineDeploymentClasses have not been deleted and are compatible.
+func ClusterClassesAreCompatible(current, desired *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if current == nil {
+		return nil
+	}
+
+	// Validate InfrastructureClusterTemplate changes desired a compatible way.
+	allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(current.Spec.Infrastructure, desired.Spec.Infrastructure)...)
+
+	// Validate control plane changes desired a compatible way.
+	allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(current.Spec.ControlPlane.LocalObjectTemplate, desired.Spec.ControlPlane.LocalObjectTemplate)...)
+	if desired.Spec.ControlPlane.MachineInfrastructure != nil && current.Spec.ControlPlane.MachineInfrastructure != nil {
+		allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(*current.Spec.ControlPlane.MachineInfrastructure, *desired.Spec.ControlPlane.MachineInfrastructure)...)
+	}
+
+	// Validate changes to MachineDeployments.
+	allErrs = append(allErrs, MachineDeploymentClassesAreCompatible(current, desired)...)
+
+	return allErrs
+}
+
+// MachineDeploymentClassesAreCompatible checks if each MachineDeploymentClass in the new ClusterClass is a compatible change from the previous ClusterClass.
+// It checks if:
+// 1) Any MachineDeploymentClass has been removed.
+// 2) If the MachineDeploymentClass.Template.Infrastructure reference has changed its Group or Kind.
+func MachineDeploymentClassesAreCompatible(current, desired *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Ensure no MachineDeployment class was removed.
+	classes := classNamesFromWorkerClass(desired.Spec.Workers)
+	for _, oldClass := range current.Spec.Workers.MachineDeployments {
+		if !classes.Has(oldClass.Class) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments"),
+					desired.Spec.Workers.MachineDeployments,
+					fmt.Sprintf("The %q MachineDeployment class can't be removed.", oldClass.Class),
+				),
```

Comment:
Does the index add value to the error output here? I would have thought the class name is a better identifier for people looking to troubleshoot, even if it results in a path that's not 100 percent correct. 

It doesn't hurt to add it though!

Reply:
I usually would prefer using the index (and not the class or topology name) as otherwise it looks like a map[string]something.

I'm not sure it's useful in this case as it would be the index of the old MDClass and I think usually the index (and the whole path) refers to an array element in the new object (aka the object the user is trying to deploy). Or phrased differently the index we would use is not the index of the invalid array element of the resource the user is deploying, so it can be misleading.

Reply:
My comment was more about there is this option, let's use it where it makes sense given that we are dealing with arrays.
Fine to have exceptions where it doesn't makes sense